File: src/models/frequency/elt/_table.py
# Copyright (c) 2022-2026 MKM Research Labs.
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

"""Building an event loss table from a catalogue (MKM-EF-001, Stage 6).

The catalogue already carries everything an ELT needs bar the losses: the event
identifiers, the per-event sampling weights, and the coverage that scales the
catchment rate down to the events the catalogue can represent. Attach the
subject's per-event losses and the table is complete.

Each event's annual rate is ``lambda_effective * weight``, where
``lambda_effective = lambda_per_year * coverage``. The weights sum to one, so
the rates sum to ``lambda_effective`` — the arrival rate of
catalogue-representable events, the same rate the loss sampler draws at. This
alignment is deliberate: it is what makes the table's closed-form average annual
loss the exact expectation of the sampler's mean aggregate loss.
"""

import numpy as np

from ..datastructures import EventCatalogue, EventLossTable


def build_event_loss_table(
    catalogue: EventCatalogue,
    event_losses: np.ndarray,
    lambda_per_year: float,
    subject_id: str,
) -> EventLossTable:
    """Assemble the event loss table for one subject.

    Args:
        catalogue: the event catalogue the losses were computed against; it
            supplies the event identifiers, the sampling weights and the
            coverage.
        event_losses: one loss per catalogue event, aligned with
            ``catalogue.event_ids``.
        lambda_per_year: the catchment arrival rate over all qualifying events;
            scaled by the catalogue coverage to the rate of representable events.
        subject_id: the gauge, property or portfolio the losses are for.

    Returns:
        An ``EventLossTable`` whose rates sum to
        ``lambda_per_year * coverage``.

    Raises:
        ValueError: if the loss vector does not align with the catalogue, which
            would silently misattribute losses to events.
    """
    losses = np.asarray(event_losses, dtype=float)
    if losses.size != catalogue.n_events:
        raise ValueError(
            f"event_losses has {losses.size} entries but the catalogue has "
            f"{catalogue.n_events} events; they must align one to one")

    lambda_effective = max(0.0, lambda_per_year) * catalogue.coverage
    rates = lambda_effective * np.asarray(catalogue.weights, dtype=float)

    return EventLossTable(
        subject_id=subject_id,
        event_ids=catalogue.event_ids,
        rates=rates,
        losses=losses,
        lambda_effective=lambda_effective,
    )

File: src/static/js/trading/stress/table.js
// Copyright (c) 2022-2026 MKM Research Labs.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

            function _tdRenderStressTable(data) {
                var wrap = document.getElementById('td-stress-table-wrap');
                if (!wrap) return;

                var trades = data.trades || [];
                var hourly = data.hourly || [];
                var summary = data.summary || {};
                var peakHour = summary.peak_p_flood_hour || 0;
                var peakData = hourly[peakHour] || {};
                var perTrade = peakData.per_trade || [];

                var ptMap = {};
                perTrade.forEach(function(pt) { ptMap[pt.swap_id] = pt; });

                var stFmtGBP = function(v) {
                    var abs = Math.abs(v);
                    var s = abs >= 1e6 ? (abs/1e6).toFixed(1) + 'M' :
                            abs >= 1e3 ? (abs/1e3).toFixed(1) + 'K' :
                            abs.toFixed(0);
                    return (v < 0 ? '-' : '') + '\u00A3' + s;
                };
                var stPnlColor = function(v) { return v >= 0 ? 'var(--green-dark)' : 'var(--red-dark)'; };

                var html = '<table style="width:100%;border-collapse:collapse;font-size:var(--size-xxs);">' +
                    '<thead><tr style="background:var(--sunken);border-bottom:2px solid var(--line-strong);">' +
                    '<th style="padding:var(--space-2) var(--space-3);text-align:left;">Swap</th>' +
                    '<th style="padding:var(--space-2) var(--space-2);text-align:center;">Dir</th>' +
                    '<th style="padding:var(--space-2) var(--space-3);text-align:right;">Notional</th>' +
                    '<th style="padding:var(--space-2) var(--space-3);text-align:right;">MTM</th>' +
                    '<th style="padding:var(--space-2) var(--space-3);text-align:right;">Cash Price</th>' +
                    '<th style="padding:var(--space-2) var(--space-3);text-align:right;">Stress P&L</th>' +
                    '<th style="padding:var(--space-2) var(--space-2);text-align:center;">Status</th>' +
                    '</tr></thead><tbody>';

                var totalNotional = 0, totalMtm = 0, totalCash = 0, totalStress = 0;
                var numTriggered = 0;

                trades.forEach(function(t) {
                    var pt = ptMap[t.swap_id] || {};
                    var cashPrice = pt.cash_price || 0;
                    var stressPnl = pt.stress_pnl || 0;
                    var dir = t.is_payer ? 'Pay' : 'Rcv';
                    var dirColor = t.is_payer ? 'var(--red-dark)' : 'var(--green-dark)';
                    var isTriggered = t.triggered_hour != null;
                    if (isTriggered) numTriggered++;
                    var statusLabel = isTriggered ?
                        '<span style="background:var(--red-dark);color:var(--panel);padding:var(--space-hair) var(--space-2);border-radius:var(--radius-sm);font-size:var(--size-8);font-weight:700;">KO H' + t.triggered_hour + '</span>' :
                        '<span style="color:var(--muted-2);font-size:var(--size-8);">Live</span>';

                    totalNotional += t.notional;
                    totalMtm += t.mtm;
                    totalCash += cashPrice;
                    totalStress += stressPnl;

                    html += '<tr style="border-bottom:1px solid var(--code);' +
                            (isTriggered ? 'background:var(--rv-bad-bg);' : '') + '">' +
                        '<td style="padding:var(--space-2) var(--space-3);font-family:monospace;font-size:var(--size-xxs);">' +
                            t.swap_id.substring(0, 12) + '</td>' +
                        '<td style="padding:var(--space-2) var(--space-2);text-align:center;color:' + dirColor + ';font-weight:600;">' +
                            dir + '</td>' +
                        '<td style="padding:var(--space-2) var(--space-3);text-align:right;">' + stFmtGBP(t.notional) + '</td>' +
                        '<td style="padding:var(--space-2) var(--space-3);text-align:right;color:' + stPnlColor(t.mtm) + ';">' +
                            stFmtGBP(t.mtm) + '</td>' +
                        '<td style="padding:var(--space-2) var(--space-3);text-align:right;color:' + stPnlColor(cashPrice) + ';">' +
                            stFmtGBP(cashPrice) + '</td>' +
                        '<td style="padding:var(--space-2) var(--space-3);text-align:right;font-weight:700;color:' +
                            stPnlColor(stressPnl) + ';">' + stFmtGBP(stressPnl) + '</td>' +
                        '<td style="padding:var(--space-2) var(--space-2);text-align:center;">' + statusLabel + '</td>' +
                        '</tr>';
                });

                html += '<tr style="border-top:2px solid var(--text);background:var(--wash);font-weight:700;">' +
                    '<td style="padding:var(--space-2) var(--space-3);" colspan="2">TOTAL</td>' +
                    '<td style="padding:var(--space-2) var(--space-3);text-align:right;">' + stFmtGBP(totalNotional) + '</td>' +
                    '<td style="padding:var(--space-2) var(--space-3);text-align:right;color:' + stPnlColor(totalMtm) + ';">' +
                        stFmtGBP(totalMtm) + '</td>' +
                    '<td style="padding:var(--space-2) var(--space-3);text-align:right;color:' + stPnlColor(totalCash) + ';">' +
                        stFmtGBP(totalCash) + '</td>' +
                    '<td style="padding:var(--space-2) var(--space-3);text-align:right;color:' + stPnlColor(totalStress) + ';">' +
                        stFmtGBP(totalStress) + '</td>' +
                    '<td style="padding:var(--space-2) var(--space-2);text-align:center;font-size:var(--size-xxs);color:var(--red-dark);">' +
                        (numTriggered > 0 ? numTriggered + '/' + trades.length : '') + '</td>' +
                    '</tr>';

                html += '</tbody></table>';

                var headerHtml = '<div style="padding:var(--space-2) 0 var(--space-4) 0;font-size:var(--size-xxs);color:var(--text-3);">' +
                    '<b>Values at peak P(flood) \u2014 Hour ' + peakHour + '</b>' +
                    ' | Water: ' + (peakData.water_level || 0).toFixed(2) + 'm' +
                    ' | P(flood): ' + ((peakData.p_flood || 0) * 100).toFixed(1) + '%' +
                    '</div>';

                wrap.innerHTML = headerHtml + html;
            }
